stdlib/public/core/Integers.swift.gyb

```diff
@@ -3232,257 +3183,496 @@ public struct DoubleWidth<T : FixedWidthInteger>
-    -> (partialValue: DoubleWidth<T>, overflow: ArithmeticOverflow) {
-    fatalError()
+    return (quotientAndRemainder(dividingBy: other).remainder, .none)
   }
 
-  public func multipliedFullWidth(by other: DoubleWidth<T>)
-    -> (high: DoubleWidth<T>, low: DoubleWidth<T>.Magnitude) {
-    fatalError()
+  public func multipliedFullWidth(by other: DoubleWidth)
+    -> (high: DoubleWidth, low: DoubleWidth.Magnitude) {
+    let isNegative = DoubleWidth.isSigned && (self < DoubleWidth()) != (other < DoubleWidth())
+
+    func mul(_ x: Low, _ y: Low) -> (partial: Low, carry: Low) {
+      let (high, low) = x.multipliedFullWidth(by: y)
+      return (low, high)
+    }
+        
+    func sum(_ x: Low, _ y: Low, _ z: Low) -> (partial: Low, carry: Low) {
+      let (sum1, overflow1) = x.addingReportingOverflow(y)
+      let (sum2, overflow2) = sum1.addingReportingOverflow(z)
+      let carry: Low = (overflow1 == .overflow ? Low(1) : Low(0))
+        + (overflow2 == .overflow ? Low(1) : Low(0))
+      return (sum2, carry)
+    }
+        
+    let lhs = self.magnitude
+    let rhs = other.magnitude
+        
+    let a = mul(rhs._storage.low, lhs._storage.low)
+    let b = mul(rhs._storage.low, lhs._storage.high)
+    let c = mul(rhs._storage.high, lhs._storage.low)
+    let d = mul(rhs._storage.high, lhs._storage.high)
+        
+    let mid1 = sum(a.carry, b.partial, c.partial)
+    let mid2 = sum(b.carry, c.carry, d.partial)
+        
+    let low = DoubleWidth<Low>((mid1.partial, a.partial))
+    let high = DoubleWidth((High(mid2.carry + d.carry), mid1.carry + mid2.partial))
+        
+    if isNegative {
+      let (lowComplement, overflow) = (~low).addingReportingOverflow(1)
+      return (~high + (overflow == .overflow ? 1 : 0), lowComplement)
+    } else {
+      return (high, low)
+    }
   }
 
   public func dividingFullWidth(
-    _ dividend: (high: DoubleWidth<T>, low: DoubleWidth<T>.Magnitude)
-  ) -> (quotient: DoubleWidth<T>, remainder: DoubleWidth<T>) {
-    fatalError()
+    _ dividend: (high: DoubleWidth, low: DoubleWidth.Magnitude)
+  ) -> (quotient: DoubleWidth, remainder: DoubleWidth) {
+    let lhs = DoubleWidth<DoubleWidth<Base>>(dividend)
+    let rhs = DoubleWidth<DoubleWidth<Base>>(self)
+    let (quotient, remainder) = lhs.quotientAndRemainder(dividingBy: rhs)
+
+    // FIXME(integers): check for overflow of quotient and remainder
+    return (DoubleWidth(quotient.low), DoubleWidth(remainder.low))
   }
 
-% for x in binaryBitwise + maskingShifts:
+% for x in binaryBitwise:
   public static func ${x.operator}=(
-    lhs: inout DoubleWidth<T>, rhs: DoubleWidth<T>
+    lhs: inout DoubleWidth, rhs: DoubleWidth
   ) {
-    fatalError()
+    lhs._storage.low ${x.operator}= rhs._storage.low
+    lhs._storage.high ${x.operator}= rhs._storage.high
   }
 % end
 
+  public static func <<=(lhs: inout DoubleWidth, rhs: DoubleWidth) {
+    if rhs < DoubleWidth((0, 0)) {
```

Comment:
`DoubleWidth` should also conform to `ExpressibleByIntegerLiteral` right? So this can be replaced by just `rhs < (0 as DoubleWidth)`? Not a big win in readability, but at least no `((...))` ;-)